src/hotspot/os_cpu/linux_aarch64/atomic_linux_aarch64.S

```diff
@@ -102,119 +130,165 @@ aarch64_atomic_cmpxchg_1_default_impl:
         stxr    w8, w2, [x0]
         cbnz    w8, 0b
-1:      mov     w0, w3
-        dmb     ish
+#endif
+1:      dmb     ish
+        mov     w0, w3
         ret
 
         .globl aarch64_atomic_cmpxchg_8_default_impl
         .align 5
 aarch64_atomic_cmpxchg_8_default_impl:
+#ifdef __ARM_FEATURE_ATOMICS
+        mov     x3, x1
+        casal   x3, x2, [x0]
+#else
         dmb     ish
         prfm    pstl1strm, [x0]
 0:      ldxr    x3, [x0]
         cmp     x3, x1
         b.ne    1f
         stxr    w8, x2, [x0]
         cbnz    w8, 0b
-1:      mov     x0, x3
-        dmb     ish
+#endif
+1:      dmb     ish
+        mov     x0, x3
         ret
 
         .globl aarch64_atomic_cmpxchg_4_release_default_impl
         .align 5
 aarch64_atomic_cmpxchg_4_release_default_impl:
+#ifdef __ARM_FEATURE_ATOMICS
+        mov     x3, x1
+        casl    w3, w2, [x0]
+#else
         prfm    pstl1strm, [x0]
 0:      ldxr    w3, [x0]
         cmp     w3, w1
         b.ne    1f
         stlxr   w8, w2, [x0]
         cbnz    w8, 0b
+#endif
 1:      mov     w0, w3
         ret
 
         .globl aarch64_atomic_cmpxchg_8_release_default_impl
         .align 5
 aarch64_atomic_cmpxchg_8_release_default_impl:
+#ifdef __ARM_FEATURE_ATOMICS
+        mov     x3, x1
+        casl    x3, x2, [x0]
+#else
         prfm    pstl1strm, [x0]
 0:      ldxr    x3, [x0]
         cmp     x3, x1
         b.ne    1f
         stlxr   w8, x2, [x0]
         cbnz    w8, 0b
+#endif
 1:      mov     x0, x3
         ret
 
         .globl aarch64_atomic_cmpxchg_4_seq_cst_default_impl
         .align 5
 aarch64_atomic_cmpxchg_4_seq_cst_default_impl:
+#ifdef __ARM_FEATURE_ATOMICS
+        mov     x3, x1
+        casal   w3, w2, [x0]
+#else
         prfm    pstl1strm, [x0]
 0:      ldaxr   w3, [x0]
         cmp     w3, w1
         b.ne    1f
         stlxr   w8, w2, [x0]
         cbnz    w8, 0b
+#endif
 1:      mov     w0, w3
         ret
 
         .globl aarch64_atomic_cmpxchg_8_seq_cst_default_impl
         .align 5
 aarch64_atomic_cmpxchg_8_seq_cst_default_impl:
+#ifdef __ARM_FEATURE_ATOMICS
+        mov     x3, x1
+        casal   x3, x2, [x0]
+#else
         prfm    pstl1strm, [x0]
 0:      ldaxr   x3, [x0]
         cmp     x3, x1
         b.ne    1f
         stlxr   w8, x2, [x0]
         cbnz    w8, 0b
+#endif
 1:      mov     x0, x3
         ret
 
 .globl aarch64_atomic_cmpxchg_1_relaxed_default_impl
         .align 5
 aarch64_atomic_cmpxchg_1_relaxed_default_impl:
+#ifdef __ARM_FEATURE_ATOMICS
+        mov     x3, x1
+        casb    w3, w2, [x0]
+#else
         prfm    pstl1strm, [x0]
 0:      ldxrb   w3, [x0]
         eor     w8, w3, w1
         tst     x8, #0xff
         b.ne    1f
         stxrb   w8, w2, [x0]
         cbnz    w8, 0b
+#endif
 1:      mov     w0, w3
         ret
 
         .globl aarch64_atomic_cmpxchg_4_relaxed_default_impl
         .align 5
 aarch64_atomic_cmpxchg_4_relaxed_default_impl:
+#ifdef __ARM_FEATURE_ATOMICS
         prfm    pstl1strm, [x0]
```

Comment:
This still looks wrong. You do want PRFM before LDXR, surely. Not before CAS.

Reply:
Thanks, fixed both points to be similar to others.